### PR TITLE
Access HTTP Body from Web Actions

### PR DESCRIPTION
**Pull request by dubeejw**
- Provide `__ow_meta_body` argument that allows access to the content of the HTTP body passed to a Web Action only when the `x-ow-raw: true` HTTP header is provided.

Close: https://github.com/openwhisk/openwhisk/issues/1947

### CODE REVIEW

**Comment by dubeejw** [comment]
@csantanapr

File: core/controller/src/main/scala/whisk/core/controller/Meta.scala [hunk 1]
```diff
@@ -310,6 +313,11 @@ trait WhiskMetaApi
         Identity.get(authStore, namespace)
     }
 
+    val validateWhiskRaw: Directive1[String] = {
+        val value = headerValueByName("x-ow-raw")
```
Comment (csantanapr) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: I think @rabbah proposed the header name is `X-OW-RAW-HTTP` instead
Also don't repeat your self, create a val with the key and don't hard code the values in multiple places, allow to update in one single place.

File: core/controller/src/main/scala/whisk/core/controller/Meta.scala [hunk 2]
```diff
@@ -330,6 +338,10 @@ trait WhiskMetaApi
                     body => process(body)
                 } ~ entity(as[FormData]) {
                     form => process(Some(form.fields.toMap.toJson.asJsObject))
+                } ~ validateWhiskRaw { headerValue =>
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: this is not the intended behavior for the header:

```
if (header is set) {
    skip entity extraction and pass body as is
} else {
    apply existing entity extraction (accept json or form data inly)
}
```

**Comment by rabbah** [comment]
@dubeejw when you add a change like this - please add a test with it (or mark it wip with a comment).

File: core/controller/src/main/scala/whisk/core/controller/Meta.scala [hunk 3]
```diff
@@ -324,12 +324,20 @@ trait WhiskMetaApi
...
 
+        val isRawHTTPEnabled = """(?i)(true)""".r
         extract(_.request.entity.data.length) { length =>
             validateSize(isWhithinRange(length))(transid) {
-                entity(as[Option[JsObject]]) {
-                    body => process(body)
-                } ~ entity(as[FormData]) {
-                    form => process(Some(form.fields.toMap.toJson.asJsObject))
+                optionalHeaderValueByName("x-ow-raw-http") {
+                    case Some(isRawHTTPEnabled(value)) =>
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: Is this equivalent to `case Some(headerValue) if headerValue.toBoolean =>`. Using Regex here feels weird.

Comment (dubeejw) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: Thanks!

File: tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala [hunk 4]
```diff
@@ -114,4 +114,56 @@ class WskWebActionsTests
...
+
+            assetHelper.withCleaner(wsk.action, name) {
+                (action, _) =>
+                    action.create(name, file, annotations = Map("web-export" -> true.toJson))
+            }
+
+            val host = getServiceURL()
+            val url = host + s"/api/v1/experimental/web/$namespace/default/webaction.text"
+            val response = RestAssured.given().header("x-ow-raw-http", "false").contentType("text/html").body(bodyContent).config(sslconfig).post(url)
+            response.statusCode shouldBe 401
```
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: The 401 status code is a defect fixed here: https://github.com/openwhisk/openwhisk/pull/1939. Test will have to be updated once that PR is merged.

File: tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala [hunk 5]
```diff
@@ -114,4 +114,56 @@ class WskWebActionsTests
             authorizedResponse.statusCode shouldBe 200
             authorizedResponse.body().asString() shouldBe namespace
     }
+
+    it should "fail to invoke web action with an unsupported content type when x-ow-raw-http header is not supplied" in withAssetCleaner(wskprops) {
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/cli/test/WskWebActionsTests.scala]: I think unit tests would be sufficient here, no? No need for a full system test.

**Comment by dubeejw** [comment]
PG2 1191, approved.

**Comment by rabbah** [comment]
superseded by #2004.

**Pull request closed without merging**